Have createNewBlock() return a BlockTemplate interface

Sjors opened the pull request:
Suggested in https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2225337100

An external program that uses the Mining interface may need quick access to some information in the block template, while it can wait a bit longer for the full raw transaction data.

This would be the case for a Stratum v2 Template Provider which needs to send a [NewTemplate](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#72-newtemplate-server---client) message message (which doesn't include transactions) as quickly as possible. It does not include the serialized block transactions.

Sjors commented:
In https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2227392678 @TheBlueMatt wrote:

> Any IPC thing that requires serializing all the transactions for all work is going to be too slow. Bitcoin Core needs to be able to track the templates it gave out and keep them around (to keep the Transaction `shared_ptr`s in a block ready to go) until they time out.

This PR returns a `BlockTemplate` interface which causes the node to hold on to the template and its block.

The goal is to be able to quickly send `NewTemplate` and after that to copy all transactions to the external application so it can respond to `RequestTransactionData` and `SubmitSolution`.

I assume your concern is that this would make `SubmitSolution` too slow because the external application needs to pass the whole block back via IPC? Or are you even worried that the TP wouldn't have the full block ready by the time `RequestTransactionData` comes in?

I could make `submitSolution` a method on the `BlockTemplate` interface. This would allow the external application to hold on to the `BlockTemplate` interface all the way until a solution comes in. This does move the burden of template memory management to the node. 

Would it need a way to unilaterally drop templates, or to tell the external application to pick which templates to drop?

Benchmarks would be useful here.

Sjors commented:
Pushed a commit that adds `submitSolution` to the `BlockTemplate` interface.

Sjors commented:
Also added `getBlockHeader()` which the Template Provider needs for the `NewTemplate` message and a few other places.

Sjors commented:
Also added `getCoinbaseTx()` and `getWitnessCommitmentIndex()` which are needed for the `NewTemplate` message.

Since this PR now contains Stratum v2 specific stuff anyway, I pulled in #30441. Still in the process of incorporating all this into #29432 to make sure I didn't miss anything.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -933,15 +936,15 @@ static RPCHelpMan getblocktemplate()
     result.pushKV("vbrequired", int(0));
 
-    result.pushKV("previousblockhash", pblock->hashPrevBlock.GetHex());
+    result.pushKV("previousblockhash", block_template->getBlockHeader().hashPrevBlock.GetHex());
```
In commit "Have createNewBlock return BlockTemplate interface" (4af3d9a483e4107301f0f0a42da781dfa79acb82)

Should be possible to avoid getBlockHeader calls by just using the `block` variable:

```diff
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -941,7 +941,7 @@ static RPCHelpMan getblocktemplate()
     result.pushKV("vbavailable", std::move(vbavailable));
     result.pushKV("vbrequired", int(0));
 
-    result.pushKV("previousblockhash", block_template->getBlockHeader().hashPrevBlock.GetHex());
+    result.pushKV("previousblockhash", block.hashPrevBlock.GetHex());
     result.pushKV("transactions", std::move(transactions));
     result.pushKV("coinbaseaux", std::move(aux));
     result.pushKV("coinbasevalue", (int64_t)block.vtx[0]->vout[0].nValue);
@@ -963,8 +963,8 @@ static RPCHelpMan getblocktemplate()
     if (!fPreSegWit) {
         result.pushKV("weightlimit", (int64_t)MAX_BLOCK_WEIGHT);
     }
-    result.pushKV("curtime", block_template->getBlockHeader().GetBlockTime());
-    result.pushKV("bits", strprintf("%08x", block_template->getBlockHeader().nBits));
+    result.pushKV("curtime", block.GetBlockTime());
+    result.pushKV("bits", strprintf("%08x", block.nBits));
     result.pushKV("height", (int64_t)(pindexPrev->nHeight+1));
 
     if (consensusParams.signet_blocks) {
```

Sjors replied:
I only used `getBlockHeader()` here to test the implementation, but it's indeed not really useful in this context.

ryanofsky reviewed (approved):
Light code review ACK 408c11e9bced08c51a7645a2de2c39c18e4360d9. This looks good, but I didn't review the getCoinbaseMerklePath implementation closely to check that it is computing the right thing.

Sjors commented:
Rebased on the changes to #30356 and extracted `GetCoinBaseMerklePath` (that'll have to be its own PR though).

ryanofsky commented on src/consensus/merkle.cpp:
```diff
@@ -83,3 +85,55 @@ uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated)
     return ComputeMerkleRoot(std::move(leaves), mutated);
 }
 
+uint256 HashTwoTxIDs(const uint256& txid1, const uint256& txid2) {
```
In commit "Add GetCoinBaseMerklePath helper" (c230cc1a72dc4e4079cb2f7b05586e6f32e844ab)

May want to declare this static to avoid creating an unused link symbol.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -838,6 +839,52 @@ class ChainImpl : public Chain
+        m_block_header = m_block_template->block.GetBlockHeader();
+    }
+
+    std::unique_ptr<CBlockTemplate> m_block_template;
+    CBlockHeader m_block_header;
```
In commit "Have createNewBlock return BlockTemplate interface" (b4782696a8ec5e21a41db3702d4b4dcc90ea40dd)

Could you move these member declarations to the bottom of the class? It makes it easier to see what state the interface accesses when state is declared at beginning or end of the class, not in the middle mixed with method definitions. 

I'm also wondering if it might be more efficient to avoid the call to CBlock::GetBlockHeader():

```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -843,14 +843,10 @@ class BlockTemplateImpl : public BlockTemplate
 {
 public:
     explicit BlockTemplateImpl(std::unique_ptr<CBlockTemplate> block_template) : m_block_template(std::move(block_template)) {
-        m_block_header = m_block_template->block.GetBlockHeader();
     }
 
-    std::unique_ptr<CBlockTemplate> m_block_template;
-    CBlockHeader m_block_header;
-
     CBlockHeader getBlockHeader() override {
-        return m_block_header;
+        return Assert(m_block_template)->block;
     }
 
     CBlock getBlock() override
@@ -883,6 +879,7 @@ public:
         return GetWitnessCommitmentIndex(m_block_template->block);
     }
 
+    std::unique_ptr<CBlockTemplate> m_block_template;
 };
 
 class MinerImpl : public Mining
```

Sjors replied:
Do you mean return `Assert(m_block_template)->block.header`?

The reason I made it a member was so that the interface would return it upon constructing `BlockTemplate`. This would then save multiple round trips over the interface later when we need a bunch of details from the header.

ryanofsky replied:
> Do you mean return `Assert(m_block_template)->block.GetBlockHeader()`?

No I think the diff above should work. There should be no need for calling GetBlockHeader() at all, much less caching the result of the call, because CBlock inherits from CBlockHeader.

I'm not suggesting dropping the getBlockHeader interface method, because that method is useful for avoiding round trips like you suggest.

Sjors replied:
> because CBlock inherits from CBlockHeader

Ah ok

> that method is useful for avoiding round trips

The choice is between having the caller be careful to avoid round trips, vs not needing round trips at all because we already got the header at construction time.

That's likely a tiny difference, so whatever is better code is probably best.

Sjors replied:
I took the patch.

ryanofsky reviewed (approved):
Code review ACK 1e8f33ec15bc468385b2807e42d60c841c407681

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/30356 landed.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -160,12 +160,12 @@ static UniValue generateBlocks(ChainstateManager& chainman, Mining& miner, const
-        std::unique_ptr<CBlockTemplate> pblocktemplate(miner.createNewBlock(coinbase_script));
-        if (!pblocktemplate.get())
+        std::unique_ptr<BlockTemplate> block_template(miner.createNewBlock(coinbase_script));
+        if (!block_template.get())
```
In commit "Have createNewBlock return BlockTemplate interface" (6536b345a346670b8733e5372f988452727a2554)

I don't see any way this condition could be true, now or previously. Would suggest changing this to assert(block_template) or CHECK_NONFATAL(block_template)

Same comment also applies to createNewBlock calls below on lines 374 and 817.

The check [here](https://github.com/bitcoin/bitcoin/blob/20ccb30b7af1c30cece2b0579ba88aa101583f07/src/node/miner.cpp#L108-L111) also appears to be dead code. I think it would be good to remove, or replace with an assert, maybe in a separate commit.

Sjors replied:
The only way this _should_ fail is `coinbase_script` is invalid (e.g. too large). But that currently results in a `throw std::runtime_error` inside `BlockAssembler::CreateNewBlock`.

Changed the three lines to `CHECK_NONFATAL`.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -838,6 +839,49 @@ class ChainImpl : public Chain
+class BlockTemplateImpl : public BlockTemplate
+{
+public:
+    explicit BlockTemplateImpl(std::unique_ptr<CBlockTemplate> block_template) : m_block_template(std::move(block_template)) {}
```
In commit "Have createNewBlock return BlockTemplate interface" (6536b345a346670b8733e5372f988452727a2554)

Would be good to guard against undefined behavior by adding `assert(m_block_template);` in the constructor body and declaring `m_block_template` as `const unique_ptr` to prevent it being reset to null.

If this is done, the `Assert` in `getBlockHeader` could also be removed so it is more consistent with the other methods.

Sjors replied:
Done

ryanofsky reviewed (approved):
Light code review ACK 52c80dff8e26c359d13c4b33e3c75a8519cdcee7, since I still didn't look very closely at merkle path function yet

Sjors committed:
Drop unneeded nullptr check from CreateNewBlock()

Sjors commented:
I split `getCoinbaseMerklePath()` and `submitSolution() out of this PR and moved them to [TODO].

Since it's astronomically unlikely multiprocess will be in the upcoming v28.0 release, we can add these functions later. They both use the new `BlockMerkleRoot` which I need to study in more detail myself.

Rebased and addressed comments above: https://github.com/bitcoin/bitcoin/pull/30440#pullrequestreview-2186655119

Marked as no longer draft.

ryanofsky commented on src/rpc/mining.cpp:
```diff
@@ -129,7 +129,7 @@ static RPCHelpMan getnetworkhashps()
 }
 
-static bool GenerateBlock(ChainstateManager& chainman, Mining& miner, CBlock& block, uint64_t& max_tries, std::shared_ptr<const CBlock>& block_out, bool process_new_block)
+static bool GenerateBlock(ChainstateManager& chainman, Mining& miner, CBlock&& block, uint64_t& max_tries, std::shared_ptr<const CBlock>& block_out, bool process_new_block)
```
In commit "Have createNewBlock return BlockTemplate interface" (f434c8562281dfec1bab37e7232ca35fd759e25f)

While changing this it would make sense to add std::move() on line 148 below so the block data does not need to be copied there.

Sjors replied:
Changed

ryanofsky reviewed (approved):
Code review ACK bde6c7f7b126beea1359990e012c76c5cafc34a6. Left one suggestion, but this change is much simpler now and look good.

ryanofsky commented on src/interfaces/mining.h:
```diff
@@ -5,26 +5,44 @@
 namespace interfaces {
 
+//! Block template interface
+class BlockTemplate
```
In commit "Have createNewBlock return BlockTemplate interface" (977ab92f8040cd139fc4b969baeec39ef58c00af)

Want to note that at some point, depending on how this interface is used, it might be a good idea to consolidate individual accessor methods into more general ones that return structs or tuples to return multiple values at the same time.

For example, if getTxFees() and getTxSigops() are typically called one after another over IPC it could make sense to combine them into a `getTxInfo()` function to save a round trip.

Sjors replied:
I'm not sure if we end using `getTxFees()` and `getTxSigops()` in stratum v2, or only for the existing stratum v1 RPC calls. Once sv2 code is a bit more mature it's indeed worth going over the interface once more.

ryanofsky reviewed (approved):
Code review ACK f3b6e728b9e9b16c71849ad044d3ef3e8240be79. Only change since the last review was adding std::move

AngusP commented on src/rpc/mining.cpp:
```diff
@@ -814,20 +812,19 @@ static RPCHelpMan getblocktemplate()
     }
     CHECK_NONFATAL(pindexPrev);
-    CBlock* pblock = &pblocktemplate->block; // pointer for convenience
+    CBlock block{block_template->getBlock()};
```
Likely a dumb question, but is switching from a `CBlock` pointer to using `getBlock` which returns a value directly going to have much of a performance impact (lots of copying)?

ryanofsky replied:
> Likely a dumb question, but is switching from a `CBlock` pointer to using `getBlock` which returns a value directly going to have much of a performance impact (lots of copying)?

Definitely not a dumb question, this changes result in an extra copy (only a move is happening on this line, but there is a copy inside the `getBlock()` function, but a copy is also removed later on due to adding std::move in `GenerateBlock`, so the net number of copies should be the same.

Sjors replied:
Marking as unresolved so we don't lose track of it.

ryanofsky replied:
Sorry, I didn't look at where GenerateBlock function was being called, I just saw one copy being added and another being removed and assumed net number of copies was the same. Agree that cost of copying should not be significant relative to other things.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -838,6 +839,51 @@ class ChainImpl : public Chain
+        return GetWitnessCommitmentIndex(m_block_template->block);
+    }
+
+    const std::unique_ptr<CBlockTemplate> m_block_template;
```
> Does this need to be public?

Convention in `interface.cpp` files is to make class members public since the classes are private and meant to be a minimal as possible (ideally just containing accessors and forwarding methods). Comment at top of the file mentions this:

https://github.com/bitcoin/bitcoin/blob/5d280130446d57d653c749005a2e363265d87686/src/node/interfaces.cpp#L82-L83

Sjors commented:
> I split `getCoinbaseMerklePath()` and `submitSolution()` out of this PR and moved them to [Sjors#53](https://github.com/Sjors/bitcoin/pull/53).

I've since `getCoinbaseMerklePath` in that PR to reuse `BlockMerkleBranch` instead of bringing its own `GetCoinBaseMerklePath` implementation. I'll move/reopen https://github.com/Sjors/bitcoin/pull/53 to the main repo after this PR is merged. I think it makes sense to keep it separate, because it requires a bit more mining specific knowledge to review than this PR.

sedited reviewed (approved):
ACK f3b6e728b9e9b16c71849ad044d3ef3e8240be79

Sjors committed:
Have createNewBlock return BlockTemplate interface

An external program that uses the Mining interface may need quick access to some information in the block template, while it can wait a bit longer for the full raw transaction data.

This would be the case for a Stratum v2 Template Provider which needs to send a NewTemplate message (which doesn't include transactions) as quickly as possible.

Sjors commented:
Rebased, ran `clang-format-diff` and updated includes and forward declarations based on https://github.com/bitcoin/bitcoin/pull/30440#discussion_r1729875816

Except that I kept `#include <node/types.h>` around because with `namespace node { struct BlockCreateOptions; }` my clang 15.0.0 compiler would complain: 

```
 initialization of incomplete type 'const node::BlockCreateOptions'
    virtual std::unique_ptr<BlockTemplate> createNewBlock(const CScript& script_pub_key, const node::BlockCreateOptions& options = {}) = 0;
```

ryanofsky reviewed (approved):
Code review ACK 33f5ab32b2651a18c9d56bef1e5b3c69dd26e199. Only changes since last review were include changes and reformatting changes.

I have to say I am not a fan of running clang-format-diff in same commits that introduce real code changes and making unrelated changes to surrounding lines of code. It particularly adds a lot of noise to this PR in the ThresholdState switch statement and makes it harder to review. IMO it is preferable to use clang-format-diff only to format code that is actually changing, not surrounding code, or to run it and commit the changes separately.

Sjors commented:
@ryanofsky I used the incantation suggested here: https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#clang-format-diffpy

I assumed this would only edit lines impacted by the commit. But I see it touched a few other places in `rpc/mining.cpp`. I removed unrelated changes from the commit.

ryanofsky reviewed (approved):
Code review ACK 23374a22f442939802c0296d6447e0db02d513da. Just reined in clang-tidy since last review so there are fewer changes

itornaza reviewed (approved):
Concept ACK 23374a22f442939802c0296d6447e0db02d513da

Do you consider updating for cmake? It would be helpful for me to build and run tests locally.

Sjors commented:
Rebased for cmake (no changes).

ryanofsky reviewed (approved):
Code review ACK a93c171faa7b4426823466e972c8f24260918bbf. Since last review, just rebased with no changes or conflicts

itornaza reviewed (approved):
Code review ACK a93c171faa7b4426823466e972c8f24260918bbf

Offering an option for the Template Provider to get block header data as soon as possible is a great addition.

Code changes look straight forward. Also, built the pr and successfully run all tests on it locally.

sedited reviewed (approved):
Re-ACK a93c171faa7b4426823466e972c8f24260918bbf

achow101 commented:
ACK a93c171faa7b4426823466e972c8f24260918bbf

Sjors commented:
@achow101 see #30510
